pkg/ottl/parser_collection.go

```diff
@@ -0,0 +1,289 @@
+	GetStatements() []string
+}
+
+// ottlParserWrapper wraps an ottl.Parser using reflection, so it can invoke exported
+// methods without knowing its generic type (transform context).
+type ottlParserWrapper[S StatementsGetter] struct {
+	parser                         reflect.Value
+	prependContextToStatementPaths func(context string, statement string) (string, error)
+}
+
+func newParserWrapper[K any, S StatementsGetter](parser *Parser[K]) *ottlParserWrapper[S] {
+	return &ottlParserWrapper[S]{
+		parser:                         reflect.ValueOf(parser),
+		prependContextToStatementPaths: parser.prependContextToStatementPaths,
+	}
+}
+
+func (g *ottlParserWrapper[S]) parseStatements(statements []string) (reflect.Value, error) {
+	method := g.parser.MethodByName("ParseStatements")
+	parseStatementsRes := method.Call([]reflect.Value{reflect.ValueOf(statements)})
+	err := parseStatementsRes[1]
+	if !err.IsNil() {
+		return reflect.Value{}, err.Interface().(error)
+	}
+	return parseStatementsRes[0], nil
+}
+
+func (g *ottlParserWrapper[S]) prependContextToStatementsPaths(context string, statements []string) ([]string, error) {
+	result := make([]string, 0, len(statements))
+	for _, s := range statements {
+		prependedStatement, err := g.prependContextToStatementPaths(context, s)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, prependedStatement)
+	}
+	return result, nil
+}
+
+// statementsConverterWrapper is reflection-based wrapper to the ParsedStatementConverter function,
+// which does not require knowing all generic parameters to be called.
+type statementsConverterWrapper[S StatementsGetter] reflect.Value
+
+func newStatementsConverterWrapper[K any, S StatementsGetter, R any](converter ParsedStatementConverter[K, S, R]) statementsConverterWrapper[S] {
+	return statementsConverterWrapper[S](reflect.ValueOf(converter))
+}
+
+func (s statementsConverterWrapper[S]) call(
+	parserCollection reflect.Value,
+	ottlParser *ottlParserWrapper[S],
+	context string,
+	statements S,
+	parsedStatements reflect.Value,
+) (reflect.Value, error) {
+	result := reflect.Value(s).Call([]reflect.Value{
+		parserCollection,
+		ottlParser.parser,
+		reflect.ValueOf(context),
+		reflect.ValueOf(statements),
+		parsedStatements,
+	})
+
+	resultValue := result[0]
+	resultError := result[1]
+	if !resultError.IsNil() {
+		return reflect.Value{}, resultError.Interface().(error)
+	}
+
+	return resultValue, nil
+}
+
+// parserCollectionParser holds an ottlParserWrapper and its respectively
+// statementsConverter function.
+type parserCollectionParser[S StatementsGetter] struct {
+	ottlParser          *ottlParserWrapper[S]
+	statementsConverter statementsConverterWrapper[S]
+}
+
+// ParserCollection is a configurable set of ottl.Parser that can handle multiple OTTL contexts
+// parsings, inferring the context and choosing the right parser for the given statements.
+type ParserCollection[S StatementsGetter, R any] struct {
```

Comment:
Is `S StatementsGetter` required? Would specifying parameters of type `StatementGetter` sufficient? Or even better can we get away with `[]string`?

Reply:
> Is S StatementsGetter required? 

Not really, I think we can drop it from the `ParserCollection`. It's only necessary for the `ParsedStatementConverter`.

Reply:
Giving it a second thought, if we drop it from `ParserCollection` and only use it on the `ParsedStatementConverter`, we wouldn't be able to enforce the `ParseStatements` argument to be the same type as the one used to add the context parser, for example: 

```go
type statementsA []string
func (s statementsA) GetStatements() []string { return s }

type statementsB []string
func (s statementsB) GetStatements() []string { return s }

func() {
  logParser, _ := ottllog.NewParser(ottlfuncs.StandardFuncs[ottllog.TransformContext](), telemetrySettings, ottllog.WithPathContextNames())

  pc, _ := ottl.NewParserCollection[any](
    componenttest.NewNopTelemetrySettings(),
    ottl.WithParserCollectionContext("log", &logParser, ottl.NewNopParsedStatementConverter[ottllog.TransformContext, statementsA]()),
  )

   // Panic as the expected type for the NewNopParsedStatementConverter function is statementsA, not statementsB
  pc.ParseStatements(statementsB([]string{`set(log.attributes["key"], log.resource.attributes["value"])`})) 
}
```

Another option would be dropping all `[S]` generic type usages replacing it by the `StatementsGetter` interface, delegating the type casting logic to the `ParsedStatementConverter` impls, example:

```go
ottl.NewParserCollection[any](
	telemetrySettings,
	ottl.WithParserCollectionContext("log", &logParser,
		func(
			collection *ottl.ParserCollection[any],
			parser *ottl.Parser[ottllog.TransformContext],
			context string,
			statements ottl.StatementsGetter, // it's currently typed statementsB
			parsedStatements []*ottl.Statement[ottllog.TransformContext]) (any, error) {
				
			b := statements.(statementsB)
                         ...
		}),
)
```